Agoda.Frameworks.Grpc/LoadBalancingCallInvoker.cs

```diff
@@ -0,0 +1,99 @@
+﻿using System;
+using Agoda.Frameworks.LoadBalancing;
+using Grpc.Core;
+
+namespace Agoda.Frameworks.Grpc
+{
+    class LoadBalancingCallInvoker : CallInvoker
+    {
+        private readonly IResourceManager<GrpcResource> _resourceManager;
+        private readonly TimeSpan? _timeout;
+        private readonly ShouldRetryPredicate _shouldRetry;
+        private readonly OnError _onError;
+
+        public LoadBalancingCallInvoker(
+            IResourceManager<GrpcResource> resourceManager,
+            TimeSpan? timeout,
+            ShouldRetryPredicate shouldRetry,
+            OnError onError)
+        {
+            _resourceManager = resourceManager;
+            _timeout = timeout;
+            _shouldRetry = shouldRetry;
+            _onError = onError;
+        }
+
+        /// <summary>
+        /// Invokes a simple remote call in a blocking fashion.
+        /// </summary>
+        public override TResponse BlockingUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request)
+        {
+            return _resourceManager.ExecuteAction((grpcResource, retryCount) =>
+            {
+                var overriddenOptions = OverrideCallOptions(options);
+                var call = new CallInvocationDetails<TRequest, TResponse>(grpcResource.Channel, method, host, overriddenOptions);
+                return Calls.BlockingUnaryCall(call, request);
+            }, _shouldRetry, _onError);
+        }
+
+        /// <summary>
+        /// Invokes a simple remote call asynchronously.
+        /// </summary>
+        public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request)
+        {
+            var callTask = _resourceManager.ExecuteAsync(async (grpcResource, retryCount) =>
+            {
+                var overriddenOptions = OverrideCallOptions(options);
+                var call = new CallInvocationDetails<TRequest, TResponse>(grpcResource.Channel, method, host, overriddenOptions);
+                var asyncCall = Calls.AsyncUnaryCall(call, request);
+                
+                await asyncCall.ResponseAsync.ConfigureAwait(false);
```

Comment:
While it seems working, it is not actually the right implementation. ResponseAsync is not the only Task from AsyncUnaryCall. Additionally, by getting `.Result` later, it involved `Wait` and obviously made the exception become `AggregateException`. 

I think in this case, you want to use `ExecuteAction` exactly like the blocking one above, except you returns Calls.AsyncUnaryCall instead. Please try it and check it's non-blocking in the test.

Notes:
https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Core/DefaultCallInvoker.cs
https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Core.Api/AsyncUnaryCall.cs

Reply:
I just realized in that case you don't get retry and it will always success.

It seems like the best way is to create your own AsyncUnaryCall and return it here. Instantiate AsyncUnaryCall object and provide wrapper methods that wraps your original calls.

Let me know if you find it hard to do.

Reply:
Nice catch! Fixed :)